src/midi/midi_soundcanvas.cpp

```diff
@@ -0,0 +1,729 @@
+{
+	const auto now_ms = PIC_FullIndex();
+
+	// Wake up the channel and update the last rendered time datum.
+	assert(mixer_channel);
+	if (mixer_channel->WakeUp()) {
+		last_rendered_ms = now_ms;
+		return 0;
+	}
+	if (last_rendered_ms >= now_ms) {
+		return 0;
+	}
+
+	// Return the number of audio frames needed to get current again
+	assert(ms_per_audio_frame > 0.0);
+
+	const auto elapsed_ms = now_ms - last_rendered_ms;
+	const auto num_audio_frames = iround(ceil(elapsed_ms / ms_per_audio_frame));
+	last_rendered_ms += (num_audio_frames * ms_per_audio_frame);
+
+	return num_audio_frames;
+}
+
+// The request to play the channel message is placed in the MIDI work FIFO
+void MidiDeviceSoundCanvas::SendMidiMessage(const MidiMessage& msg)
+{
+	std::vector<uint8_t> message(msg.data.begin(), msg.data.end());
+
+	MidiWork work{std::move(message),
+	              GetNumPendingAudioFrames(),
+	              MessageType::Channel};
+
+	work_fifo.Enqueue(std::move(work));
+}
+
+// The request to play the sysex message is placed in the MIDI work FIFO
+void MidiDeviceSoundCanvas::SendSysExMessage(uint8_t* sysex, size_t len)
+{
+	std::vector<uint8_t> message(sysex, sysex + len);
+	MidiWork work{std::move(message), GetNumPendingAudioFrames(), MessageType::SysEx};
+	work_fifo.Enqueue(std::move(work));
+}
+
+// The callback operates at the audio frame-level, steadily adding samples to
+// the mixer until the requested numbers of audio frames is met.
+void MidiDeviceSoundCanvas::MixerCallback(const int requested_audio_frames)
+{
+	assert(mixer_channel);
+
+	// Report buffer underruns
+	constexpr auto warning_percent = 5.0f;
+
+	if (const auto percent_full = audio_frame_fifo.GetPercentFull();
+	    percent_full < warning_percent) {
+		static auto iteration = 0;
+		if (iteration++ % 100 == 0) {
+			LOG_WARNING("SOUNDCANVAS: Audio buffer underrun");
+		}
+		had_underruns = true;
+	}
+
+	static std::vector<AudioFrame> audio_frames = {};
+
+	const auto has_dequeued = audio_frame_fifo.BulkDequeue(audio_frames,
+	                                                       requested_audio_frames);
+
+	if (has_dequeued) {
+		mixer_channel->AddSamples_sfloat(requested_audio_frames,
+		                                 &audio_frames[0][0]);
+
+		last_rendered_ms = PIC_FullIndex();
+	} else {
+		assert(!audio_frame_fifo.IsRunning());
+		mixer_channel->AddSilence();
+	}
+}
+
+void MidiDeviceSoundCanvas::RenderAudioFramesToFifo(const int num_audio_frames)
+{
+	assert(num_audio_frames > 0);
+
+	static std::vector<float> left  = {};
+	static std::vector<float> right = {};
+
+	// Maybe expand the vectors
+	if (check_cast<int>(left.size()) < num_audio_frames) {
+		left.resize(num_audio_frames);
+		right.resize(num_audio_frames);
+	}
+
+	float* audio_out[] = {left.data(), right.data()};
+
+	clap.plugin->Process(audio_out, num_audio_frames, clap.event_list);
+	clap.event_list.Clear();
+
+	for (auto i = 0; i < num_audio_frames; ++i) {
+		audio_frame_fifo.Enqueue({left[i], right[i]});
+	}
```

Comment:
Potential optimization here (not sure how much of a difference it will make in practice so completely optional):

Interleave the channels into a single vector here and then you can use `BulkEnqueue()` so you're not taking and releasing a mutex lock for every audio frame.

Reply:
Seems worthwhile to do, yeah.

Reply:
Well I guess if the bulk of the calls through here are for a single frame anyway, you might as well leave it as is.  It's kind of a tradeoff anyway as you'd be making an extra copy if you interleave them so you might accidentally make it slower :laughing: 

Anyway, I doubt it will make much a difference either way.  The very large majority of time in this function is spent doing the SC55 emulation.

Reply:
On a second thought, I think you're right. I kinda thought the same thing what you said when I initially wrote this (that it doesn't really matter). Less code == better when it doesn't make a significant positive difference, so leaving it.